File: src/java/main/helper/RichJsonFileHelper.java
package helper;

import com.fasterxml.jackson.databind.ObjectMapper;
import core.RichJsonCommandHolder;
import other.RichJsonConfig;
import core.RichJsonParser;

import java.io.File;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;

public class RichJsonFileHelper {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final Map<String, Object> FILE_CACHE = new HashMap<>();

    /**
     * Liest ein Verzeichnis wie eine JSON-Datei und löst RichJson auf.
     */
    public static Map<String, Object> readDirectory(String pathStr, boolean executeLateApplies) {
        File folder = new File(pathStr);
        File[] entries = folder.listFiles();
        Map<String, Object> rv = new HashMap<>();

        if (entries == null) return rv;

        if (!executeLateApplies) {
            RichJsonCommandHolder.lateApplies.forEach(cmd -> RichJsonCommandHolder.setCommandEnabled(cmd, false));
        }

        for (File entry : entries) {
            String name = entry.getName();
            if (entry.isFile()) {
                String nameWithoutExtension = name.contains(".") ? name.substring(0, name.lastIndexOf('.')) : name;
                rv.put(nameWithoutExtension, readFile(entry.getPath(), true));
            } else if (entry.isDirectory()) {
                rv.put(name, readDirectory(entry.getPath(), true));
            }
        }

        if (!executeLateApplies) {
            RichJsonCommandHolder.lateApplies.forEach(cmd -> RichJsonCommandHolder.setCommandEnabled(cmd, true));
        }

        return rv;
    }

    /**
     * Liest eine JSON-Datei und führt den RichJsonParser darauf aus.
     */
    public static Object readFile(String pathStr, boolean executeLateApplies) {
        // 1. Cache Check
        if (RichJsonConfig.fileCacheEnabled && FILE_CACHE.containsKey(pathStr)) {
            return FILE_CACHE.get(pathStr);
        }

        if (RichJsonConfig.fileCacheEnabled) {
            FILE_CACHE.put(pathStr, new HashMap<String, Object>());
        }

        // 2. Late Applies Management
        if (!executeLateApplies) {
            RichJsonCommandHolder.lateApplies.forEach(cmd -> RichJsonCommandHolder.setCommandEnabled(cmd, false));
        }

        Object rv;
        try {
            Path path = Paths.get(pathStr);
            String content = Files.readString(path, StandardCharsets.UTF_8);

            rv = MAPPER.readValue(content, Object.class);

            RichJsonParser parser = new RichJsonParser();
            rv = parser.parse(rv, true);

        } catch (Exception e) {
            throw new RuntimeException("Error reading RichJson file: " + pathStr, e);
        }

        // 3. Reaktivierung
        if (!executeLateApplies) {
            RichJsonCommandHolder.lateApplies.forEach(cmd -> RichJsonCommandHolder.setCommandEnabled(cmd, true));
        }

        // 4. Cache Update
        if (RichJsonConfig.fileCacheEnabled) {
            Object cached = FILE_CACHE.get(pathStr);
            if (cached instanceof Map && rv instanceof Map) {
                RichJsonHelper.mergeIntoTarget((Map<String, Object>) cached, (Map<String, Object>) rv);
            } else {
                FILE_CACHE.put(pathStr, rv);
            }
        }

        return rv;
    }
}

File: src/c#/main/helper/RichJsonFileHelper.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using Newtonsoft.Json;
using main.core;
using main.other;

namespace main.helper
{
    public class RichJsonFileHelper
    {
        private static readonly Dictionary<string, object> FILE_CACHE = new Dictionary<string, object>();

        /// <summary>
        /// Reads a directory like a JSON file and resolves RichJson.
        /// </summary>
        public static Dictionary<string, object> ReadDirectory(string pathStr, bool executeLateApplies)
        {
            var rv = new Dictionary<string, object>();

            if (!Directory.Exists(pathStr)) return rv;

            string[] entries = Directory.GetFileSystemEntries(pathStr);

            if (!executeLateApplies)
            {
                RichJsonCommandHolder.LateApplies.ForEach(cmd => RichJsonCommandHolder.SetCommandEnabled(cmd, false));
            }

            foreach (string entry in entries)
            {
                string name = Path.GetFileName(entry);
                if (File.Exists(entry))
                {
                    string nameWithoutExtension = Path.GetFileNameWithoutExtension(entry);
                    rv[nameWithoutExtension] = ReadFile(entry, true);
                }
                else if (Directory.Exists(entry))
                {
                    rv[name] = ReadDirectory(entry, true);
                }
            }

            if (!executeLateApplies)
            {
                RichJsonCommandHolder.LateApplies.ForEach(cmd => RichJsonCommandHolder.SetCommandEnabled(cmd, true));
            }

            return rv;
        }

        /// <summary>
        /// Reads a JSON file and executes the RichJsonParser on it.
        /// </summary>
        public static object ReadFile(string pathStr, bool executeLateApplies)
        {
            // 1. Cache Check
            if (RichJsonConfig.FileCacheEnabled && FILE_CACHE.ContainsKey(pathStr))
            {
                return FILE_CACHE[pathStr];
            }

            if (RichJsonConfig.FileCacheEnabled)
            {
                FILE_CACHE[pathStr] = new Dictionary<string, object>();
            }

            // 2. Late Applies Management
            if (!executeLateApplies)
            {
                RichJsonCommandHolder.LateApplies.ForEach(cmd => RichJsonCommandHolder.SetCommandEnabled(cmd, false));
            }

            object rv;
            try
            {
                string content = File.ReadAllText(pathStr, Encoding.UTF8);

                // Using Newtonsoft.Json to perfectly unpack the types deep down
                rv = JsonConvert.DeserializeObject<object>(content);

                RichJsonParser parser = new RichJsonParser();
                rv = parser.Parse(rv, true);
            }
            catch (Exception e)
            {
                throw new Exception("Error reading RichJson file: " + pathStr, e);
            }

            // 3. Reactivation
            if (!executeLateApplies)
            {
                RichJsonCommandHolder.LateApplies.ForEach(cmd => RichJsonCommandHolder.SetCommandEnabled(cmd, true));
            }

            // 4. Cache Update
            if (RichJsonConfig.FileCacheEnabled)
            {
                object cached = FILE_CACHE[pathStr];
                if (cached is IDictionary<string, object> cachedMap && rv is IDictionary<string, object> rvMap)
                {
                    RichJsonHelper.MergeIntoTarget(cachedMap, rvMap);
                }
                else
                {
                    FILE_CACHE[pathStr] = rv;
                }
            }

            return rv;
        }
    }
}